In [1]:

import numpy as np
import pandas as pd
import random 

import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

plt.rcParams["figure.figsize"] = (10,8) 

/home/sanfermen/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,
/home/sanfermen/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,


In [2]:
#Cargamos el dataframe con el que vamos a trabajar.
df = pd.read_csv("../ficheros/diamons_estandarizados.csv", index_col = 0)
df.head(2)

carat      cut color clarity     depth     table  price         x  \
1 -0.777778    Ideal     E     SI2 -1.428571  1.333333    326 -0.965909   
2 -0.746032  Premium     E     SI1 -3.500000  2.666667    326 -0.875000   

          y         z  
1 -1.000000 -1.018182  
2 -0.867816 -1.018182

In [3]:
lm = ols("price ~ carat + cut + color + clarity + depth + table + x + y + z", data= df).fit()
sm.stats.anova_lm(lm)

df        sum_sq       mean_sq             F         PR(>F)
cut           4.0  5.236258e+09  1.309065e+09    340.536548  1.894856e-289
color         6.0  8.943270e+09  1.490545e+09    387.746370   0.000000e+00
clarity       7.0  9.785157e+09  1.397880e+09    363.640665   0.000000e+00
carat         1.0  1.714838e+11  1.714838e+11  44609.332964   0.000000e+00
depth         1.0  2.320500e+08  2.320500e+08     60.364878   8.041345e-15
table         1.0  6.246957e+08  6.246957e+08    162.506678   3.699683e-37
x             1.0  4.784580e+09  4.784580e+09   1244.647929  4.324508e-269
y             1.0  3.505944e+07  3.505944e+07      9.120268   2.529298e-03
z             1.0  1.721697e+05  1.721697e+05      0.044788   8.323957e-01
Residual  46578.0  1.790516e+11  3.844123e+06           NaN            NaN

Según vemos aquí, todas nuestras variables tienen un p-valor menor de 0.05, por lo que debemos rechazar la hipótesis nula, lo que quiere decir que todas las variables influyen en nuestra variable respuesta, que es el precio de los diamantes.

In [4]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.529
Model:                            OLS   Adj. R-squared:                  0.529
Method:                 Least Squares   F-statistic:                     2275.
Date:                Mon, 23 Jan 2023   Prob (F-statistic):               0.00
Time:                        09:42:23   Log-Likelihood:            -4.1940e+05
No. Observations:               46602   AIC:                         8.389e+05
Df Residuals:                   46578   BIC:                         8.391e+05
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         2836.6933     90.759     31.255      0.000    2658.805    3014.582
cut[T.Good]       -149.3547     62.232     -2.400      0.016    -271.330     -27.379
cut[T.Ideal]      -237.8308     56.904     -4.179      0.000    -349.364    -126.298
cut[T.Premium]      30.3994     57.535      0.528      0.597     -82.370     143.169
cut[T.Very Good]  -103.1000     58.083     -1.775      0.076    -216.944      10.744
color[T.E]         -15.5140     32.903     -0.472      0.637     -80.003      48.975
color[T.F]         222.9021     33.424      6.669      0.000     157.391     288.413
color[T.G]         328.0472     32.547     10.079      0.000     264.255     391.839
color[T.H]         330.6912     34.537      9.575      0.000     262.998     398.385
color[T.I]         488.5503     39.321     12.425      0.000     411.480     565.621
color[T.J]         535.4601     47.817     11.198      0.000     441.738     629.183
clarity[T.IF]     -314.8988     93.256     -3.377      0.001    -497.682    -132.115
clarity[T.SI1]     -49.2080     80.424     -0.612      0.541    -206.840     108.424
clarity[T.SI2]     155.4675     81.257      1.913      0.056      -3.797     314.732
clarity[T.VS1]     -76.0185     81.763     -0.930      0.353    -236.276      84.239
clarity[T.VS2]     -49.1752     80.720     -0.609      0.542    -207.388     109.038
clarity[T.VVS1]   -394.9185     85.906     -4.597      0.000    -563.295    -226.542
clarity[T.VVS2]   -198.6200     84.063     -2.363      0.018    -363.384     -33.856
carat             6083.9532     80.800     75.297      0.000    5925.584    6242.322
depth             -211.4403     10.968    -19.278      0.000    -232.938    -189.943
table             -178.5429     13.043    -13.688      0.000    -204.108    -152.978
x                -3413.8343    149.477    -22.839      0.000   -3706.811   -3120.858
y                  345.6558    114.250      3.025      0.002     121.723     569.588
z                  -14.4777     68.410     -0.212      0.832    -148.562     119.607
==============================================================================
Omnibus:                    10905.187   Durbin-Watson:                   0.366
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           108228.223
Skew:                           0.843   Prob(JB):                         0.00
Kurtosis:                      10.273   Cond. No.                         32.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Nuestro *R cuadrado* es 0.529, que no es lo mejor, pero indica que sí que afectan a la variable respuesta.

- Respecto al *error estándar*, los que tienen el valor más bajo son **Depth** y **Table**, lo que quiere decir que la estimación de la variable respuesta será más precisa. El **color** también parece bastante preciso. Sin embargo, tanto x como y tienen valores muy altos.

- Respecto a la *P*, la mayoría son menores de 0.05, con lo cual son importantes al hacer las predicciones. 
    Respecto al c**orte(cut)**, la categoría *premium*, pero nos quedaríamos con **cut** igual porque el resto de las categorías sí son importantes.

    Con el **color** pasa lo mismo, que todas son menores que 0.05, excepto el **color** T.E, por lo que también nos quedaríamos con la variable **color**.

    Con **clarity** tendremos 3 menores que 0.05, y el resto de las categorías serían mayores, pero con que haya una que sí influya ya no nos podríamos deshacer de la variable.

    **Carat**, **depth**, **table**, **x** e **y** tienen valores menores de 0.05, por lo que sí son importantes al hacer la predicción.

    La **z** en este caso parece que no es significativa para hacer las predicciones, cosa que nos ha sorprendido porque no era lo que parecía en los anteriores análisis. Pero según esto, podríamos eliminar esta variable.

- Respecto al *coeficiente*, vemos que sobre todo el **carat (quilates)** tiene un coeficiente muy alto. Cuanto mayor son los quilates, el precio de los diamantes sube mucho. En el caso de **depth** y **table** la relación es inversamente proporcional.

In [5]:
# Eliminamos la columna z y guardamos otro csv para no perder los datos con lo que empezamos hoy, ya que no tenemos claro cual seguiremos usando
df.drop(columns=["z"], axis= 1, inplace= True)

In [6]:
df.head(1)

carat    cut color clarity     depth     table  price         x    y
1 -0.777778  Ideal     E     SI2 -1.428571  1.333333    326 -0.965909 -1.0

In [7]:
df.to_csv("../ficheros/diamons_estand_sinz.csv")

Hacemos lo mismo con los datos sin estandarizar, para comprobar los resultados

In [9]:
df2= pd.read_csv("../ficheros/diamons_01.csv", index_col = 0)
df2.head(2)

carat      cut color clarity  depth  table  price     x     y     z
1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [10]:
lm2 = ols("price ~ carat + cut + color + clarity + depth + table + x + y + z", data= df2).fit()
sm.stats.anova_lm(lm2)

df        sum_sq       mean_sq              F         PR(>F)
cut           4.0  4.790534e+09  1.197634e+09    1908.550474   0.000000e+00
color         6.0  8.518113e+09  1.419685e+09    2262.412529   0.000000e+00
clarity       7.0  9.270288e+09  1.324327e+09    2110.448921   0.000000e+00
carat         1.0  3.300181e+11  3.300181e+11  525917.282722   0.000000e+00
depth         1.0  9.667681e+05  9.667681e+05       1.540643   2.145279e-01
table         1.0  3.251085e+07  3.251085e+07      51.809340   6.201355e-13
x             1.0  3.195148e+08  3.195148e+08     509.179193  3.462949e-112
y             1.0  4.681036e+06  4.681036e+06       7.459705   6.311735e-03
z             1.0  9.547653e+05  9.547653e+05       1.521515   2.173961e-01
Residual  50247.0  3.153047e+10  6.275095e+05            NaN            NaN

In [11]:
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.918
Model:                            OLS   Adj. R-squared:                  0.918
Method:                 Least Squares   F-statistic:                 2.446e+04
Date:                Mon, 23 Jan 2023   Prob (F-statistic):               0.00
Time:                        12:16:19   Log-Likelihood:            -4.0687e+05
No. Observations:               50271   AIC:                         8.138e+05
Df Residuals:                   50247   BIC:                         8.140e+05
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept        -1294.5174    309.367     -4.184      0.000   -1900.879    -688.155
cut[T.Good]        483.8604     24.403     19.828      0.000     436.029     531.691
cut[T.Ideal]       691.2052     24.321     28.420      0.000     643.535     738.876
cut[T.Premium]     635.0755     23.437     27.097      0.000     589.138     681.013
cut[T.Very Good]   607.4155     23.489     25.860      0.000     561.377     653.454
color[T.E]        -162.9624     12.813    -12.719      0.000    -188.076    -137.849
color[T.F]        -212.0875     13.010    -16.302      0.000    -237.586    -186.588
color[T.G]        -346.8630     12.769    -27.164      0.000    -371.891    -321.835
color[T.H]        -700.0175     13.642    -51.313      0.000    -726.756    -673.279
color[T.I]       -1104.2423     15.486    -71.307      0.000   -1134.595   -1073.890
color[T.J]       -1741.8227     19.338    -90.073      0.000   -1779.725   -1703.920
clarity[T.IF]     4119.4940     36.735    112.142      0.000    4047.494    4191.494
clarity[T.SI1]    2900.6629     31.286     92.714      0.000    2839.342    2961.984
clarity[T.SI2]    2148.8808     31.414     68.406      0.000    2087.310    2210.452
clarity[T.VS1]    3638.1765     31.971    113.795      0.000    3575.512    3700.840
clarity[T.VS2]    3369.7774     31.475    107.063      0.000    3308.087    3431.468
clarity[T.VVS1]   3969.8691     33.800    117.452      0.000    3903.621    4036.117
clarity[T.VVS2]   3982.3313     32.876    121.133      0.000    3917.895    4046.768
carat             8841.9856     51.946    170.214      0.000    8740.170    8943.801
depth              -29.8565      3.361     -8.884      0.000     -36.443     -23.270
table              -16.8614      2.126     -7.931      0.000     -21.029     -12.694
x                 -502.8348     34.526    -14.564      0.000    -570.505    -435.164
y                   76.2371     26.838      2.841      0.005      23.635     128.840
z                  -31.0028     25.134     -1.233      0.217     -80.266      18.260
==============================================================================
Omnibus:                    12827.160   Durbin-Watson:                   1.072
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           205461.271
Skew:                           0.801   Prob(JB):                         0.00
Kurtosis:                      12.774   Cond. No.                     7.46e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.46e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

La columna **z** sigue sin afectar a nuestra variable resultado, por lo que vamos a eliminarla.

In [12]:
df2.drop(columns=["z"], axis= 1, inplace= True)

In [13]:
df2.head(1)

carat    cut color clarity  depth  table  price     x     y
1   0.23  Ideal     E     SI2   61.5   55.0    326  3.95  3.98

In [14]:
df2.to_csv("../ficheros/diamons_sinz.csv")